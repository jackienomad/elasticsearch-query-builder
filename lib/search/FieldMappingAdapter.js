export class FieldMappingAdapter {
  /*------------------------------CONSTRUCTOR------------------------------*/
  constructor(fieldMapping) {
    if (!fieldMapping) throw new Error('fieldMapping is required')
    this.fieldMapping = fieldMapping
  }

  destroy() {
    this.fieldMapping = null
  }

  /*------------------------------METHODS------------------------------*/
  getMappedField(selectedField, isUntouched = false) {
    if (!this.fieldMapping[selectedField]) return ''

    if (isUntouched) {
      return (
        this.fieldMapping[selectedField].not_analyzed_field ||
        this.fieldMapping[selectedField].field ||
        ''
      )
    } else {
      return (
        this.fieldMapping[selectedField].field ||
        this.fieldMapping[selectedField].not_analyzed_field ||
        ''
      )
    }
  }

  getMiscField(miscField) {
    if (!this.fieldMapping[miscField]) return ''
    const correspondingField = this.fieldMapping[miscField].field

    if (!this.fieldMapping[correspondingField]) return correspondingField
    return this.fieldMapping[correspondingField].field || ''
  }
}
export default FieldMappingAdapter
