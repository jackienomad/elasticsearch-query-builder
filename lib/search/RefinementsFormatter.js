export class RefinementsFormatter {
  static format(response, refinementGroups) {
    return Object.keys(response.aggregations)
      .map(field => {
        const refinementGroup = {
          field,
          category: refinementGroups[field]
        }

        if (response.aggregations[field].buckets) {
          refinementGroup.totalRefinements =
            response.aggregations[field].buckets.length
          refinementGroup.refinements = response.aggregations[
            field
          ].buckets.map(({ key: name, doc_count: count }) => ({
            name,
            count
          }))
        } else {
          // miscellaneous
          refinementGroup.totalCount =
            response.aggregations[field].doc_count || 0
        }

        return refinementGroup
      })
      .sort(({ field: field1 }, { field: field2 }) => {
        const ucField1 = field1.toUpperCase()
        const ucField2 = field2.toUpperCase()
        if (ucField1 < ucField2) {
          return -1
        } else if (ucField2 < ucField1) {
          return 1
        }

        return 0
      })
  }
}

export default RefinementsFormatter
