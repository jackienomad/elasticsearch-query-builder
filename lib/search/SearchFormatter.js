export class SearchFormatter {
  static formatResponse(esResponse) {
    return esResponse.hits.hits.map(({ _id, _source }) => ({
      ArrivalDate: _source['Arrival Date'] || '',
      BillOfLading: _source['Bill of Lading Number'] || '',
      BillTypeCode: _source['Bill Type'] || '',
      CarrierAddress: _source['Carrier']['Address'] || '',
      CarrierCity: (_source['Carrier'] && _source['Carrier']['City']) || '',
      CarrierCode: (_source['Carrier'] && _source['Carrier']['Code']) || '',
      CarrierName: (_source['Carrier'] && _source['Carrier']['Name']) || '',
      CarrierState: (_source['Carrier'] && _source['Carrier']['State']) || '',
      CarrierZip: (_source['Carrier'] && _source['Carrier']['Zip Code']) || '',
      Consignee: (_source['Consignee'] && _source['Consignee']['Name']) || '',
      ConsigneeAddress:
        (_source['Consignee'] && _source['Consignee']['Address']) || '',
      ConsigneeAddressBreakdown:
        _source['Consignee'] &&
        _source['Consignee']['Cleaned Address Breakdown'],
      ConsigneeCity:
        (_source['Consignee'] &&
          _source['Consignee']['Cleaned Address Breakdown'] &&
          _source['Consignee']['Cleaned Address Breakdown']['PlaceName']) ||
        '',
      ConsigneeId:
        (_source['Consignee'] &&
          _source['Consignee']['Cleaned Address Breakdown'] &&
          _source['Consignee']['Cleaned Address Breakdown']['_id']) ||
        '',
      ConsigneeState:
        (_source['Consignee'] &&
          _source['Consignee']['Cleaned Address Breakdown'] &&
          _source['Consignee']['Cleaned Address Breakdown']['StateName']) ||
        '',
      ConsigneeStreet:
        (_source['Consignee'] &&
          _source['Consignee']['Cleaned Address Breakdown'] &&
          _source['Consignee']['Cleaned Address Breakdown']['StreetName']) ||
        '',
      ConsigneeStreetNumber:
        (_source['Consignee'] &&
          _source['Consignee']['Cleaned Address Breakdown'] &&
          _source['Consignee']['Cleaned Address Breakdown']['AddressNumber']) ||
        '',
      ConsigneeZipCode:
        (_source['Consignee'] &&
          _source['Consignee']['Cleaned Address Breakdown'] &&
          _source['Consignee']['Cleaned Address Breakdown']['ZipCode']) ||
        '',
      ContainerHeight:
        (_source['Containers']['0'] && _source['Containers']['0']['Height']) ||
        '',
      ContainerLength:
        (_source['Containers']['0'] && _source['Containers']['0']['Length']) ||
        '',
      ContainerNumber:
        (_source['Containers']['0'] &&
          _source['Containers']['0']['Container Number']) ||
        '',
      ContainerType:
        (_source['Containers']['0'] && _source['Containers']['0']['Type']) ||
        '',
      ContainerWidth:
        (_source['Containers']['0'] && _source['Containers']['0']['Width']) ||
        '',
      Containers: _source['Containers'] || '',
      CountryOfOrigin: _source['Country of Origin'] || '',
      DestinationPort: _source['Port of Destination'] || '',
      EquipmentDescriptionCode:
        (_source['Containers']['0'] &&
          _source['Containers']['0']['Equipment Description']) ||
        '',
      ForeignPort: _source['Foreign Port of Lading'] || '',
      GrossWeightKgs: _source['Weight'] || '',
      GrossWeightLbs: _source['Weight'] || '',
      HSCode: _source['HS Code'] || '',
      InBondEntryType: _source['In-Bond Entry Type'] || '',
      LoadStatus:
        _source['Containers']['0'] && _source['Containers']['0']['Load Status'],
      ManifestNumber: _source['Manifest Number'] || '',
      MarksAndNumbers:
        (_source['Containers']['0'] && _source['Containers']['0']['Marks']) ||
        '',
      MasterBillOfLading: _source['Master Bill of Lading Number'] || '',
      Measurement: _source['Measurement'] || '',
      MeasurementUnit: _source['Measurement Unit'] || '',
      NotifyParty:
        (_source['Notify Party'] && _source['Notify Party']['Name']) || '',
      NotifyPartyAddress:
        (_source['Notify Party'] && _source['Notify Party']['Address']) || '',
      NumberOfContainers: _source['Number of Containers'] || '',
      PlaceOfReceipt: _source['Place of Receipt'] || '',
      ProductDescription:
        (_source['Containers']['0']['Cargoes'] &&
          _source['Containers']['0']['Cargoes']['Description']) ||
        '',
      ProductKeywords:
        (_source['Containers']['0']['Cargoes'] &&
          _source['Containers']['0']['Cargoes']['Keywords']) ||
        '',
      Quantity: _source['Manifest Quantity'] || '',
      QuantityUnit: _source['Manifest Unit'] || '',
      SealNumbers:
        _source['Containers']['0'] &&
        _source['Containers']['0']['Seal Numbers'],
      ShipmentId: _id || '',
      Shipper: (_source['Shipper'] && _source['Shipper']['Name']) || '',
      ShipperAddress:
        (_source['Shipper'] && _source['Shipper']['Address']) || '',
      ShipperId: (_source['Shipper'] && _source['Shipper']['_id']) || '',
      TypeOfService:
        (_source['Containers'] && _source['Containers']['Type of Service']) ||
        '',
      VesselName: _source['Vessel Name'] || '',
      VoyageNumber: _source['Voyage Number'] || ''
    }))
  }
}

export default SearchFormatter
