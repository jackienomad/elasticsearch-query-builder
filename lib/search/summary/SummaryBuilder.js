export class SummaryBuilder {
  /*------------------------------CONSTRUCTOR------------------------------*/
  constructor(fieldMapping) {
    if (typeof fieldMapping !== 'object') {
      throw new Error('fieldMapping is required')
    }

    this.fieldMapping = fieldMapping
  }

  destroy() {
    this.fieldMapping = null
  }

  /*------------------------------METHODS------------------------------*/
  buildAggregates() {
    return this.fieldMapping.countables.reduce((acc, val) => {
      acc[val.alias] = {
        sum: {
          field: this.fieldMapping[val.field].field
        }
      }

      return acc
    }, {})
  }

  buildQuery(searchQuery, index, dateRanges) {
    const body = []
    const aggregations = this.buildAggregates()

    dateRanges.forEach(({ start, end }) => {
      const query = JSON.parse(JSON.stringify(searchQuery))

      query.bool.filter[1].range['Arrival Date'].gte = parseInt(start)
      query.bool.filter[1].range['Arrival Date'].lte = parseInt(end)

      body.push({ index: `${index}` })
      body.push({
        query,
        size: 0,
        aggregations
      })
    })

    return { body }
  }
}

export default SummaryBuilder
