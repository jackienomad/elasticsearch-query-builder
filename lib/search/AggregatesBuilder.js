import { FieldMappingAdapter } from './FieldMappingAdapter'

export class AggregatesBuilder {
  static build(fields, fieldMapping) {
    const fieldAdapter =
      fieldMapping instanceof FieldMappingAdapter
        ? fieldMapping
        : new FieldMappingAdapter(fieldMapping)

    const aggregations = {}
    fields.forEach(field => {
      aggregations[field] = {
        terms: {
          field: fieldAdapter.getMappedField(field, true),
          size: 100
        }
      }
    })

    return aggregations
  }
}

export default AggregatesBuilder
