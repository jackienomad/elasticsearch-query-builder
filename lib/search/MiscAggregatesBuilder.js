import { FieldMappingAdapter } from './FieldMappingAdapter'
import { MiscTypes } from './searchUtils'

export class MiscAggregatesBuilder {
  static generateFilterObject(misc) {
    let filter
    if (misc === MiscTypes.BLANK_CONSIGNEES_EXCLUDED) {
      filter = {
        exists: {
          field: 'Consignee'
        }
      }
    } else if (misc === MiscTypes.BLANK_SHIPPERS_EXCLUDED) {
      filter = {
        exists: {
          field: 'Shipper'
        }
      }
    } else if (misc === MiscTypes.MASTER_SHIPMENTS_EXCLUDED) {
      return {
        bool: {
          must_not: {
            match_phrase: {
              Consignee: 'Master'
            }
          }
        }
      }
    } else if (misc === MiscTypes.HOUSE_SHIPMENTS_ONLY) {
      filter = {
        bool: {
          should: [
            {
              match_phrase: {
                'Bill Type': 'H'
              }
            },
            {
              match_phrase: {
                'Bill Type': 'House'
              }
            }
          ]
        }
      }
    } else if (misc === MiscTypes.MASTER_SHIPMENTS_ONLY) {
      filter = {
        bool: {
          must: {
            match_phrase: {
              'Bill Type': 'Master'
            }
          }
        }
      }
    }

    return filter
  }

  static build(fields, fieldMapping) {
    const fieldAdapter =
      fieldMapping instanceof FieldMappingAdapter
        ? fieldMapping
        : new FieldMappingAdapter(fieldMapping)

    const aggregations = {}
    fields.forEach(field => {
      aggregations[field] = {
        filter: MiscAggregatesBuilder.generateFilterObject(field)
      }
    })

    return aggregations
  }
}

export default MiscAggregatesBuilder
