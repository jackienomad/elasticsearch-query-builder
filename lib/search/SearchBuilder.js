import {
  getMinimumByType,
  Operators,
  SearchTypes,
  MiscTypes
} from './searchUtils'
import { FieldMappingAdapter } from './FieldMappingAdapter'

export class SearchBuilder {
  /*------------------------------CONSTRUCTOR------------------------------*/
  constructor(index, type, fieldMapping) {
    this.index = index
    this.type = type
    this.fieldMapping = fieldMapping
    this.fieldMappingAdapter = new FieldMappingAdapter(fieldMapping)
    this.searchQuery = null
    this.mustFilterQuery = null
    this.notQuery = null
    this.sorting = null
  }

  destroy() {
    this.fieldMappingAdapter.destroy()

    this.index = null
    this.type = null
    this.fieldMapping = null
    this.fieldMappingAdapter = null
    this.searchQuery = null
    this.mustFilterQuery = null
    this.notQuery = null
    this.sorting = null
  }

  /*------------------------------METHODS------------------------------*/
  buildKeywords(keywords) {
    if (!this.notQuery) this.notQuery = []
    if (!this.searchQuery) this.searchQuery = []

    let queryGroup
    // build keywords
    keywords.forEach(({ field, operator, keyword, type }) => {
      const mappedField = this.fieldMappingAdapter.getMappedField(field)
      if (operator === Operators.NOT) {
        this.notQuery.push({
          match_phrase: { [mappedField]: keyword }
        })
      } else {
        let queryObject
        if (type === SearchTypes.EXACT) {
          queryObject = {
            match_phrase: { [mappedField]: keyword }
          }
        } else {
          queryObject = {
            match: {
              [mappedField]: {
                query: keyword,
                minimum_should_match: getMinimumByType(type)
              }
            }
          }
        }

        // init querygroup
        if (operator === Operators.AND || !queryGroup) {
          queryGroup = {
            bool: {
              should: []
            }
          }
          this.searchQuery.push(queryGroup)
        }

        queryGroup.bool.should.push(queryObject)
      }
    })
  }

  buildInclude(include) {
    const includeReturn = {}
    include.forEach(({ field, keyword }) => {
      const includedField = this.fieldMappingAdapter.getMappedField(field, true)
      if (!includeReturn[includedField]) includeReturn[includedField] = []
      includeReturn[includedField].push({
        match_phrase: { [includedField]: keyword }
      })
    })

    return Object.keys(includeReturn).map(field => ({
      bool: {
        should: includeReturn[field]
      }
    }))
  }

  buildExclude(exclude) {
    return exclude.map(({ field, keyword }) => ({
      match_phrase: {
        [this.fieldMappingAdapter.getMappedField(field)]: keyword
      }
    }))
  }

  buildMisc(misc) {
    if (!this.notQuery) this.notQuery = []
    if (!this.mustFilterQuery) this.mustFilterQuery = []
    if (!this.searchQuery) this.searchQuery = []

    misc &&
      misc.forEach(elem => {
        const miscField = this.fieldMappingAdapter.getMiscField(elem)

        if (
          elem === MiscTypes.BLANK_CONSIGNEES_EXCLUDED ||
          elem === MiscTypes.BLANK_SHIPPERS_EXCLUDED
        ) {
          this.searchQuery.push({
            exists: {
              field: miscField
            }
          })
        } else if (elem === MiscTypes.MASTER_SHIPMENTS_EXCLUDED) {
          this.notQuery.push({
            match_phrase: { [miscField]: 'Master' }
          })
        } else if (elem === MiscTypes.HOUSE_SHIPMENTS_ONLY) {
          this.mustFilterQuery.push({
            bool: {
              should: [
                {
                  match_phrase: { [miscField]: 'H' }
                },
                {
                  match_phrase: { [miscField]: 'House' }
                }
              ]
            }
          })
        } else if (elem === MiscTypes.MASTER_SHIPMENTS_ONLY) {
          this.mustFilterQuery.push({
            match_phrase: { [miscField]: 'Master' }
          })
        }
      })

    return {
      notQuery: this.notQuery,
      mustFilterQuery: this.mustFilterQuery,
      searchQuery: this.searchQuery
    }
  }

  buildSorting(sortBy) {
    let sortField = 'Arrival Date'
    let sortOrder = 'asc'
    this.sorting = []

    if (sortBy && sortBy.length > 0) {
      if (sortBy.charAt(0) === '-') {
        sortOrder = 'desc'
        sortField = this.fieldMappingAdapter.getMappedField(
          sortBy.substr(1),
          true
        )
      } else if (sortBy.charAt(0) === '+') {
        sortOrder = 'asc'
        sortField = this.fieldMappingAdapter.getMappedField(
          sortBy.substr(1),
          true
        )
      } else {
        sortOrder = 'asc'
        sortField = this.fieldMappingAdapter.getMappedField(sortBy, true)
      }
    }

    this.sorting.push(
      {
        [sortField]: {
          order: sortOrder
        }
      },
      '_score'
    )

    return this
  }

  buildQuery(
    keywords,
    include,
    exclude,
    misc,
    sortBy,
    limit,
    page,
    arrivalDates = ''
  ) {
    let [gte = '', lte = ''] = (arrivalDates && arrivalDates.split('-')) || []
    this.searchQuery = []
    this.mustFilterQuery = []
    this.notQuery = []

    this.buildKeywords(keywords)

    if (include) this.mustFilterQuery.push(...this.buildInclude(include))
    if (exclude) this.notQuery.push(...this.buildExclude(exclude))

    this.buildMisc(misc)
    this.buildSorting(sortBy)

    const querySearch = {
      size: limit,
      from: page * limit - limit,
      track_scores: true,
      sort: this.sorting,
      query: {
        bool: {
          must: this.searchQuery,
          must_not: this.notQuery,
          filter: [
            ...this.mustFilterQuery,
            {
              range: {
                'Arrival Date': {
                  gte: gte,
                  lte: lte
                }
              }
            }
          ]
        }
      }
    }

    return {
      index: this.index,
      type: this.type,
      body: querySearch
    }
  }
}

export default SearchBuilder
