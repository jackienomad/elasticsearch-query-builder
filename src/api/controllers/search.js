import { client } from '../models/queryBuilderModel'
import { SearchBuilder } from 'qb/@/search/SearchBuilder'
import { SearchFormatter } from 'qb/@/search/SearchFormatter'

export async function fetchShipments({
  keywords,
  include,
  exclude,
  misc,
  sortBy,
  limit,
  page,
  index,
  type,
  arrivalDate = ''
}) {
  const builder = new SearchBuilder(index, type)
  const query = builder.buildQuery(
    keywords,
    include,
    exclude,
    misc,
    sortBy,
    limit,
    page,
    arrivalDate
  )

  const resp = await client.search(query)
  return {
    total: resp.hits.total,
    max_score: resp.hits.max_score,
    data: {
      items: SearchFormatter.formatResponse(resp)
    }
  }
}

export default fetchShipments
