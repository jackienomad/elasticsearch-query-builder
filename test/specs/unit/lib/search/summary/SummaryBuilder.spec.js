import fieldMapping from '../dataset/sampleFieldMapping'
import { expectedQueryBuilderReturn } from './expectedQueryBuilderReturn'
import { SummaryBuilder } from 'qb/@/search/summary/SummaryBuilder'
import { SearchBuilder } from 'qb/@/search/SearchBuilder'
import Faker from 'faker'

describe('SummaryBuilder', () => {
  describe('constructor and destructor', () => {
    it('should be instantiable', () => {
      expect(() => {
        new SummaryBuilder(fieldMapping)
      }).not.toThrowError()
    })

    it('should be destroyable', () => {
      const builder = new SummaryBuilder(fieldMapping)
      expect(() => {
        builder.destroy()
      }).not.toThrowError()
    })

    it('should throw an error on missing fieldMapping', () => {
      expect(() => {
        new SummaryBuilder()
      }).toThrowError()
    })
  })

  describe('buildAggregates', () => {
    it('should build aggregates from fieldMapping', () => {
      const builder = new SummaryBuilder(fieldMapping)
      const output = builder.buildAggregates()

      expect(output).toMatchSnapshot()
    })
  })

  describe('buildQuery', () => {
    const builder = new SummaryBuilder(fieldMapping)

    const index = 'us_import'

    const dateRanges = [
      {
        start: 1534867200,
        end: 1535471999
      },
      {
        start: 1535472000,
        end: 1536076799
      },
      {
        start: 1536076800,
        end: 1536192000
      }
    ]

    afterAll(() => {
      builder.destroy()
      builder = null
    })

    it('should build query', () => {
      const output = builder.buildQuery(
        expectedQueryBuilderReturn,
        index,
        dateRanges
      )

      expect(output).toMatchSnapshot()
    })
  })
})
