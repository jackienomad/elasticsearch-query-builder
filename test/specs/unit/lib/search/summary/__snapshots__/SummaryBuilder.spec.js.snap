// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SummaryBuilder buildAggregates should build aggregates from fieldMapping 1`] = `
Object {
  "no_of_containers": Object {
    "sum": Object {
      "field": "Number of Containers",
    },
  },
  "weight": Object {
    "sum": Object {
      "field": "Weight",
    },
  },
}
`;

exports[`SummaryBuilder buildQuery should build query 1`] = `
Object {
  "body": Array [
    Object {
      "index": "us_import",
    },
    Object {
      "aggregations": Object {
        "no_of_containers": Object {
          "sum": Object {
            "field": "Number of Containers",
          },
        },
        "weight": Object {
          "sum": Object {
            "field": "Weight",
          },
        },
      },
      "query": Object {
        "bool": Object {
          "filter": Array [
            Object {
              "bool": Object {
                "should": Array [
                  Object {
                    "match_phrase": Object {
                      "Consignee.Cleaned Address Breakdown.StateName.untouched": "CA",
                    },
                  },
                  Object {
                    "match_phrase": Object {
                      "Consignee.Cleaned Address Breakdown.StateName.untouched": "WA",
                    },
                  },
                ],
              },
            },
            Object {
              "range": Object {
                "Arrival Date": Object {
                  "gte": 1534867200,
                  "lte": 1535471999,
                },
              },
            },
          ],
          "must": Array [
            Object {
              "bool": Object {
                "should": Array [
                  Object {
                    "match": Object {
                      "Consignee.Name": Object {
                        "minimum_should_match": "75%",
                        "query": "apple",
                      },
                    },
                  },
                  Object {
                    "match": Object {
                      "Consignee.Name": Object {
                        "minimum_should_match": "75%",
                        "query": "orange",
                      },
                    },
                  },
                ],
              },
            },
            Object {
              "exists": Object {
                "field": "Consignee.Name",
              },
            },
          ],
          "must_not": Array [
            Object {
              "match_phrase": Object {
                "Consignee.Cleaned Address Breakdown.PlaceName": "IRVINE",
              },
            },
          ],
        },
      },
      "size": 0,
    },
    Object {
      "index": "us_import",
    },
    Object {
      "aggregations": Object {
        "no_of_containers": Object {
          "sum": Object {
            "field": "Number of Containers",
          },
        },
        "weight": Object {
          "sum": Object {
            "field": "Weight",
          },
        },
      },
      "query": Object {
        "bool": Object {
          "filter": Array [
            Object {
              "bool": Object {
                "should": Array [
                  Object {
                    "match_phrase": Object {
                      "Consignee.Cleaned Address Breakdown.StateName.untouched": "CA",
                    },
                  },
                  Object {
                    "match_phrase": Object {
                      "Consignee.Cleaned Address Breakdown.StateName.untouched": "WA",
                    },
                  },
                ],
              },
            },
            Object {
              "range": Object {
                "Arrival Date": Object {
                  "gte": 1535472000,
                  "lte": 1536076799,
                },
              },
            },
          ],
          "must": Array [
            Object {
              "bool": Object {
                "should": Array [
                  Object {
                    "match": Object {
                      "Consignee.Name": Object {
                        "minimum_should_match": "75%",
                        "query": "apple",
                      },
                    },
                  },
                  Object {
                    "match": Object {
                      "Consignee.Name": Object {
                        "minimum_should_match": "75%",
                        "query": "orange",
                      },
                    },
                  },
                ],
              },
            },
            Object {
              "exists": Object {
                "field": "Consignee.Name",
              },
            },
          ],
          "must_not": Array [
            Object {
              "match_phrase": Object {
                "Consignee.Cleaned Address Breakdown.PlaceName": "IRVINE",
              },
            },
          ],
        },
      },
      "size": 0,
    },
    Object {
      "index": "us_import",
    },
    Object {
      "aggregations": Object {
        "no_of_containers": Object {
          "sum": Object {
            "field": "Number of Containers",
          },
        },
        "weight": Object {
          "sum": Object {
            "field": "Weight",
          },
        },
      },
      "query": Object {
        "bool": Object {
          "filter": Array [
            Object {
              "bool": Object {
                "should": Array [
                  Object {
                    "match_phrase": Object {
                      "Consignee.Cleaned Address Breakdown.StateName.untouched": "CA",
                    },
                  },
                  Object {
                    "match_phrase": Object {
                      "Consignee.Cleaned Address Breakdown.StateName.untouched": "WA",
                    },
                  },
                ],
              },
            },
            Object {
              "range": Object {
                "Arrival Date": Object {
                  "gte": 1536076800,
                  "lte": 1536192000,
                },
              },
            },
          ],
          "must": Array [
            Object {
              "bool": Object {
                "should": Array [
                  Object {
                    "match": Object {
                      "Consignee.Name": Object {
                        "minimum_should_match": "75%",
                        "query": "apple",
                      },
                    },
                  },
                  Object {
                    "match": Object {
                      "Consignee.Name": Object {
                        "minimum_should_match": "75%",
                        "query": "orange",
                      },
                    },
                  },
                ],
              },
            },
            Object {
              "exists": Object {
                "field": "Consignee.Name",
              },
            },
          ],
          "must_not": Array [
            Object {
              "match_phrase": Object {
                "Consignee.Cleaned Address Breakdown.PlaceName": "IRVINE",
              },
            },
          ],
        },
      },
      "size": 0,
    },
  ],
}
`;
